Packages/vcs/Lib/vcsvtk/boxfillpipeline.py

```diff
@@ -0,0 +1,383 @@
+        """Overrides baseclass implementation."""
+        if self._gm.boxfill_type != "custom":
+            self._updateContourLevelsAndColorsForBoxfill()
+        else:
+            self._updateContourLevelsAndColorsForCustomBoxfill()
+
+        if isinstance(self._contourLevels, numpy.ndarray):
+            self._contourLevels = self._contourLevels.tolist()
+
+
+    def _updateContourLevelsAndColorsForBoxfill(self):
+        """Set contour information for a standard boxfill."""
+        # Compute levels
+        nlev = (self._gm.color_2 - self._gm.color_1) + 1
+        if numpy.allclose(self._gm.level_1, 1.e20) or \
+           numpy.allclose(self._gm.level_2, 1.e20):
+            self._contourLevels = vcs.mkscale(self._scalarRange[0],
+                                              self._scalarRange[1])
+            if len(self._contourLevels) == 1: # constant value ?
+                self._contourLevels = [self._contourLevels[0],
+                                       self._contourLevels[0] + .00001]
+            self._legend = vcs.mklabels(self._contourLevels)
+            dx = (self._contourLevels[-1] - self._contourLevels[0]) / nlev
+            self._contourLevels = numpy.arange(self._contourLevels[0],
+                                               self._contourLevels[-1] + dx, dx)
+        else:
+            if self._gm.boxfill_type == "log10":
+                levslbls = vcs.mkscale(numpy.ma.log10(self._gm.level_1),
+                                       numpy.ma.log10(self._gm.level_2))
+                self._contourLevels = \
+                      vcs.mkevenlevels(numpy.ma.log10(self._gm.level_1),
+                                       numpy.ma.log10(self._gm.level_2),
+                                       nlev=nlev)
+            else:
+                levslbls = vcs.mkscale(self._gm.level_1, self._gm.level_2)
+                self._contourLevels = vcs.mkevenlevels(self._gm.level_1,
+                                                       self._gm.level_2,
+                                                       nlev=nlev)
+            if len(self._contourLevels) > 25:
+                ## Too many colors/levels need to prettyfy this for legend
+                self._legend = vcs.mklabels(levslbls)
+                ## Make sure extremes are in
+                legd2 = vcs.mklabels([self._contourLevels[0],
+                                      self._contourLevels[-1]])
+                self._legend.update(legd2)
+            else:
+                self._legend = vcs.mklabels(self._contourLevels)
+            if self._gm.boxfill_type == "log10":
+                for k in self._legend.keys():
+                    self._legend[float(numpy.ma.log10(self._legend[k]))] = \
+                          self._legend[k]
+                    del(self._legend[k])
+
+        # Use consecutive colors:
+        self._contourColors = range(self._gm.color_1, self._gm.color_2 + 1)
+
+    def _updateContourLevelsAndColorsForCustomBoxfill(self):
```

Comment:
this bit is probably reusable for isofill. Shouldn't we create a "levels" class and inherit from this, so that we don't need to duplicate efforts between boxfill custon/isofill/meshfill